File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/10816-숫자 카드 2/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 500005

int N;
int A[NM];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i];
}

int lower_bound(int A[], int L, int R, int X) {
	int ans = R + 1;
	while (L <= R) {
		int mid = (L + R) / 2;
		if (A[mid] >= X) {
			ans = mid;
			R = mid - 1;
		}
		else {
			L = mid + 1;
		}
	}
	return ans;
}

int upper_bound(int A[], int L, int R, int X) {
	int ans = R + 1;
	while (L <= R) {
		int mid = (L + R) / 2;
		if (A[mid] > X) {
			ans = mid;
			R = mid - 1;
		}
		else {
			L = mid + 1;
		}
	}
	return ans;
}

#include <vector>
void pro() {
	sort(A + 1, A + 1 + N);

	int M;
	cin >> M;
	for (int i = 1; i <= M; i++) {
		int X;
		cin >> X;
		cout << upper_bound(A, 1, N, X) - lower_bound(A, 1, N, X) << ' ';
	}
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/10816-숫자 카드 2/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N;
    static int[] A;

    static void input() {
        N = scan.nextInt();
        A = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            A[i] = scan.nextInt();
        }
    }

    static int lower_bound(int[] A, int L, int R, int X) {
        int ans = R + 1;
        while (L <= R) {
            int mid = (L + R) / 2;
            if (A[mid] >= X) {
                ans = mid;
                R = mid - 1;
            } else {
                L = mid + 1;
            }
        }
        return ans;
    }

    static int upper_bound(int[] A, int L, int R, int X) {
        int ans = R + 1;
        while (L <= R) {
            int mid = (L + R) / 2;
            if (A[mid] > X) {
                ans = mid;
                R = mid - 1;
            } else {
                L = mid + 1;
            }
        }
        return ans;
    }

    static void pro() {
        int M = scan.nextInt();
        Arrays.sort(A, 1, N + 1);
        for (int i = 1; i <= M; i++) {
            int X = scan.nextInt();
            int upper = upper_bound(A, 1, N, X);
            int lower = lower_bound(A, 1, N, X);
            sb.append(upper - lower).append(' ');
        }
        System.out.println(sb);
    }

    public static void main(String[] args) {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}